sql: limit non-table decimal types in casts

madelynnblue opened the pull request:
While here change the error message to be more similar to postgres.

Fixes #15833

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15834)
<!-- Reviewable:end -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod comments. quick turnaround!

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/col_types.go, line 205 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834#-KjmeqGdmbbM880qyN5w:-KjmeqGdmbbM880qyN5x:b4redha) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d3dde1dc399c2d7ba4836bc1c662e1b6f05c77c/pkg/sql/parser/col_types.go#L205)):*
> ```Go
> 	}
> 	if scale > precision {
> 		return errors.Errorf("scale (%d) must be between 0 and precision (%d)", scale, precision)
> ```

Please use the proper pgwire errors here. If you weren't already familiar with how to determine them, open `psql` against postgres, type `\set VERBOSITY verbose`, and run statements that produce the corresponding errors. Then, look up the error code in `codes.go`.

---

*[pkg/sql/parser/col_types.go, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834#-KjmfTwiN8yTBXrCuq1C:-KjmfTwiN8yTBXrCuq1D:bf8dho6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d3dde1dc399c2d7ba4836bc1c662e1b6f05c77c/pkg/sql/parser/col_types.go#L227)):*
> ```Go
> 			lt = fmt.Sprintf("10^%d", v)
> 		}
> 		return errors.Errorf("value with precision %d, scale %d must round to an absolute value less than %s", precision, scale, lt)
> ```

ditto pgerror

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834#-:-KjmgPjQdk2I3s0_y0Bu:b-fb5xkf)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: limit non-table decimal types in casts

Change cast expressions to decimals to support precision and
scale-limited values. We already did this when storing data in a table,
but this information was ignored during decimal conversions.

While here change the error message to be more similar to postgres.

Fixes #15833

mjibson commented:
Updated with pgerror changes and more detail in the commit message.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/parser/col_types.go, line 205 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834#-KjmeqGdmbbM880qyN5w:-KjnkymUxrpFrFHtXo1f:b-5hjlvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d3dde1dc399c2d7ba4836bc1c662e1b6f05c77c/pkg/sql/parser/col_types.go#L205)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Please use the proper pgwire errors here. If you weren't already familiar with how to determine them, open `psql` against postgres, type `\set VERBOSITY verbose`, and run statements that produce the corresponding errors. Then, look up the error code in `codes.go`.
</blockquote></details>

I didn't know this was a thing. Done.

---

*[pkg/sql/parser/col_types.go, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834#-KjmfTwiN8yTBXrCuq1C:-Kjnl-Z4n3h2VbKUPfuU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d3dde1dc399c2d7ba4836bc1c662e1b6f05c77c/pkg/sql/parser/col_types.go#L227)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

ditto pgerror
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834#-:-Kjnl-xPVCFkKxwrDa7c:btfxrqn)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15834)*
<!-- Sent from Reviewable.io -->
